cmd/libsnap-confine-private/string-utils.c

```diff
@@ -69,3 +70,48 @@ int sc_must_snprintf(char *str, size_t size, const char *format, ...)
 
 	return n;
 }
+
+char *sc_must_stpcpy(char *buf, size_t buf_size, char *dest, const char *src)
+{
+	// Set errno in case we die.
+	errno = 0;
+	if (buf == NULL) {
+		die("cannot append string: buffer is NULL");
+	}
+	if (dest == NULL) {
+		die("cannot append string: destination is NULL");
+	}
+	if (src == NULL) {
+		die("cannot append string: source is NULL");
+	}
+	// Sanity check, the code doesn't need buffers larger than a few KBs so
+	// prevent corrupted or otherwise huge buffers from seeming "valid".
+	if (buf_size >= 0xFFFF) {
+		// NOTE: using %zd to format size_t as ssize_t which is more useful for
+		// -1 and similar huge values and also is better to test as it is
+		// independent of machine word size.
+		die("cannot append string: buffer size (%zd) exceeds internal limit", (ssize_t) buf_size);
+
+	}
+	// Sanity check, dest points to the inside of the buffer.
+	if (dest == &buf[buf_size]) {
+		die("cannot append string: destination points"
+		    " to the end of the buffer");
+	}
+	if (dest > &buf[buf_size]) {
+		die("cannot append string: destination points"
+		    " %td byte(s) beyond the buffer", dest - &buf[buf_size]);
+	}
+	if (dest < buf) {
+		die("cannot append string: destination points"
+		    " %td byte(s) in front of the buffer", buf - dest);
+	}
+	// Sanity check the new content fits the buffer.
+	size_t src_len = strlen(src);
+	if (&dest[src_len] >= &buf[buf_size]) {
+		die("cannot append string: buffer overflow of %td byte(s)",
+		    &dest[src_len] - &buf[buf_size] + 1);
+	}
+	memcpy(dest, src, src_len + 1);
+	return &dest[src_len + 1];
```

Comment:
I agree about how `stpcpy` is an optimization detail for how append should work. If you want I can rename this to `sc_append_string` and just use that everywhere.